engine/access/ingestion2/results_forest_test.go

```diff
@@ -0,0 +1,1684 @@
+		// protocol state (based on our test setup). In previous tests, we have confirmed that the corresponding
+		// results `s.results[0]` and `s.results[1]` are accepted.
+		forest := s.createForest()
+
+		// Test Scenario: inconsistent parent result
+		// ┏━━━━━━━━━━━┓                      ┏━━━━━━━━━━━┓
+		// ┃ Block b0  ┃ ◄━━━━━━━━━━━━━━━━━━━ ┃ Block b1  ┃
+		// ┗━━━━━━━━━━━┛                      ┗━━━━━━━━━━━┛
+		//                                         ∆
+		//                                         │
+		// ┌───────────┐         inconsistent ╔═══════════╗
+		// │ Result s0 │ ◁════-x-x-x-x-x-x-══ ║ Result r1 ║
+		// └───────────┘        parent result ╚═══════════╝
+		//        │
+		//        ∇
+		// ┏━━━━━━━━━━━┓
+		// ┃ Block a0  ┃
+		// ┗━━━━━━━━━━━┛
+
+		// Setup: construct block `a0` and add corresponding result `s0` to forest. result `s0` should become the latest sealed.
+		a0 := blockGen.Fixture(blockGen.WithParentHeader(s.initialSealedHeader), blockGen.WithView(s.initialSealedHeader.View+10))
+		s.allBlocks[a0.ID()] = a0
+		s0 := resultGen.Fixture(resultGen.WithBlock(a0), resultGen.WithPreviousResult(s.initialSealedResult))
+		pipeline1 := s.createPipeline(s0) // configure mock processing pipeline
+		pipeline1.On("SetSealed").Return().Once()
+
+		err := forest.AddSealedResult(s0)
+		s.Require().NoError(err)
+		s.RequireLatestSealedResult(forest, s0, a0.View)
+
+		// Constructing incorrect result
+		r1 := resultGen.Fixture(resultGen.WithBlock(s.blocks[1]), resultGen.WithPreviousResult(s0))
+		_ = s.createPipeline(r1) // configure mock processing pipeline; r2 should be rejected, so no state changes
+
+		// Add incorrect result
+		err = forest.AddSealedResult(r1)
+		s.Error(err)
+		s.True(forestmodule.IsInvalidVertexError(err))
+		s.requireUnknown(forest, r1.ID())
+
+		// only the latest persisted sealed result and the result first result should be in the forest
+		s.Equal(uint(2), forest.Size())
+		s.assertContainer(forest, s.initialSealedResult.ID(), ingestion2.ResultSealed)
+		s.assertContainer(forest, s0.ID(), ingestion2.ResultSealed)
+		s.RequireLatestSealedResult(forest, s0, a0.View)
+	})
+}
+
+// TestAddSealedResult_ConcurrentWithAddReceipt tests the concurrent calls to AddSealedResult and
+// AddReceipt are handled as expected.
+//
+// This test creates a main fork of sealed results and execution forks with up to 3 results
+// descending from each sealed result. Then it concurrently calls AddSealedResult for all sealed
+// results and AddReceipt for all conflicting results, and verifies the final state of the forest.
+func (s *ResultsForestSuite) TestAddSealedResult_ConcurrentWithAddReceipt() {
+	blockGen := s.g.Blocks()
+	receiptGen := s.g.ExecutionReceipts()
+
+	forest := s.createForest()
+
+	// with 50 blocks, there should be ~200 results in the forest
+	blocks := blockGen.List(50, blockGen.WithParentHeader(s.initialSealedHeader))
```

Comment:
I think we might be creating blocks with view numbers overlapping with the views from the blocks generated in `SetupTest`: https://github.com/onflow/flow-go/blob/f20da9477a3dd99502c8b3a3ace8262cdb88df4c/engine/access/ingestion2/results_forest_test.go#L74

Leo has a very pragmatic approach: he just provides a map as input to these functions (you could either use the optional argument pattern. Or you just overwrite `s.allBlocks`:

```suggestion
	// We should avoid generating conflicting blocks for the same view, because blocks are assumed to be certified in the
	// context of the ResultsForest. Among all blocks produced for a view, consensus guarantees that at most one can be certified.
	// Hence, we re-initialize the map here, representing the blocks in storage.
	s.allBlocks = make(map[flow.Identifier]*flow.Block) // initialSealedBlock is included by default by `s.headers` mock

	// with 50 blocks, there should be ~200 results in the forest
	blocks := blockGen.List(50, blockGen.WithParentHeader(s.initialSealedHeader))
```